app/forms/vpc-router-route-common.tsx

```diff
@@ -0,0 +1,117 @@
+  internetGatewayTargetValue:
+    'For ‘Internet gateway’ targets, the value must be ‘outbound’',
+  // https://github.com/oxidecomputer/omicron/blob/914f5fd7d51f9b060dcc0382a30b607e25df49b2/nexus/src/app/vpc_router.rs#L201-L204
+  noNewRoutesOnSystemRouter: 'User-provided routes cannot be added to a system router',
+  // https://github.com/oxidecomputer/omicron/blob/914f5fd7d51f9b060dcc0382a30b607e25df49b2/nexus/src/app/vpc_router.rs#L300-L304
+  noDeletingRoutesOnSystemRouter: 'System routes cannot be deleted',
+  // https://github.com/oxidecomputer/omicron/blob/914f5fd7d51f9b060dcc0382a30b607e25df49b2/nexus/src/app/vpc_router.rs#L136-L138
+  noDeletingSystemRouters: 'System routers cannot be deleted',
+}
+
+// VPCs cannot be specified as a destination in custom routers
+// https://github.com/oxidecomputer/omicron/blob/4f27433d1bca57eb02073a4ea1cd14557f70b8c7/nexus/src/app/vpc_router.rs#L363
+const destTypes: Record<Exclude<RouteDestination['type'], 'vpc'>, string> = {
+  ip: 'IP',
+  ip_net: 'IP network',
+  subnet: 'Subnet',
+}
+
+// Subnets and VPCs cannot be used as a target in custom routers
+// https://github.com/oxidecomputer/omicron/blob/4f27433d1bca57eb02073a4ea1cd14557f70b8c7/nexus/src/app/vpc_router.rs#L362-L368
+const targetTypes: Record<Exclude<RouteTarget['type'], 'subnet' | 'vpc'>, string> = {
+  ip: 'IP',
+  instance: 'Instance',
+  internet_gateway: 'Internet gateway',
+  drop: 'Drop',
+}
+
+const toItems = (mapping: Record<string, string>) =>
+  Object.entries(mapping).map(([value, label]) => ({ value, label }))
+
+type RouteFormFieldsProps = {
+  form: UseFormReturn<RouteFormValues>
+  isDisabled?: boolean
+}
+export const RouteFormFields = ({ form, isDisabled }: RouteFormFieldsProps) => {
+  const { control } = form
+  const targetType = form.watch('target.type')
+  return (
+    <>
+      {isDisabled && (
+        <Message variant="info" content={routeFormMessage.vpcSubnetNotModifiable} />
+      )}
+      <NameField name="name" control={control} disabled={isDisabled} />
+      <DescriptionField name="description" control={control} disabled={isDisabled} />
+      <ListboxField
+        name="destination.type"
+        label="Destination type"
+        control={control}
+        items={toItems(destTypes)}
+        placeholder="Select a destination type"
+        required
+        disabled={isDisabled}
+      />
+      <TextField
+        name="destination.value"
+        label="Destination value"
+        control={control}
+        placeholder="Enter a destination value"
+        required
+        disabled={isDisabled}
+      />
+      <ListboxField
+        name="target.type"
+        label="Target type"
+        control={control}
+        items={toItems(targetTypes)}
+        placeholder="Select a target type"
+        required
+        onChange={(value) => {
+          form.setValue('target.value', value === 'internet_gateway' ? 'outbound' : '')
```

Comment:
I see this is a change from the previous logic. Good one, it makes way more sense to always clear it. There is one minor issue but it's tolerable: "changing" the type to the same value it already is fires `onChange` and clears the value (I tried it). The brute force fix would be to check if it's different from the current value before firing `setValue`. I'm also fine without that. I checked the headless listbox docs and it doesn't seem like there's an obvious way to tell it not to fire `onChange` unless the value has actually changed.

Reply:
I experimented with a few iterations with the brute force "pass a conditional prop through with the frozen originalTargetType and compare it to the form's targetType at the onChange firing time", but things got fairly squirrelly. For now, I'll get this folded in, as I think clearing the value is better as the default action. Absolutely good to revisit it down the road if we find UX friction on the field being problematic.